nixos/modules/services/matrix/matrix-synapse.nix

```diff
@@ -0,0 +1,768 @@
+                            "consent"
+                            "federation"
+                            "keys"
+                            "media"
+                            "metrics"
+                            "openid"
+                            "replication"
+                            "static"
+                          ]);
+                          description = ''
+                            List of resources to host on this listener.
+                          '';
+                          example = [
+                            "client"
+                          ];
+                        };
+                        compress = mkOption {
+                          type = types.bool;
+                          description = ''
+                            Should synapse compress HTTP responses to clients that support it?
+                            This should be disabled if running synapse behind a load balancer
+                            that can do automatic compression.
+                          '';
+                        };
+                      };
+                    });
+                    description = ''
+                      List of HTTP resources to serve on this listener.
+                    '';
+                  };
+                };
+              });
+              default = [ {
+                port = 8008;
+                bind_addresses = [
+                  "127.0.0.1"
+                ];
+                type = "http";
+                tls = false;
+                x_forwarded = true;
+                resources = [ {
+                  names = [
+                    "client"
+                  ];
+                  compress = true;
+                } {
+                  names = [
+                    "federation"
+                  ];
+                  compress = false;
+                } ];
+              } ];
+              description = ''
+                List of ports that Synapse should listen on, their purpose and their configuration.
+              '';
+            };
+
+            database.name = mkOption {
+              type = types.enum [
+                "sqlite3"
+                "psycopg2"
+              ];
+              default = if versionAtLeast config.system.stateVersion "18.03"
+                then "psycopg2"
+                else "sqlite3";
+               defaultText = literalExpression ''
+                if versionAtLeast config.system.stateVersion "18.03"
+                then "psycopg2"
+                else "sqlite3"
+              '';
+              description = ''
+                The database engine name. Can be sqlite3 or psycopg2.
+              '';
+            };
+
+            database.args.database = mkOption {
+              type = types.str;
+              default = {
+                sqlite3 = "${cfg.dataDir}/homeserver.db";
+                psycopg2 = "matrix-synapse";
+              }.${cfg.settings.database.name};
```

Comment:
Wondering if reusing that same expression within `literalExpression` is fine or not because it uses `cfg.dataDir` and `cfg.settings.database.name`

```patch
@@ -490,6 +492,12 @@ in {
                 sqlite3 = "${cfg.dataDir}/homeserver.db";
                 psycopg2 = "matrix-synapse";
               }.${cfg.settings.database.name};
+              defaultText = literalExpression ''
+              {
+                sqlite3 = "${cfg.dataDir}/homeserver.db";
+                psycopg2 = "matrix-synapse";
+              }.${cfg.settings.database.name};
+              '';
               description = ''
                 Name of the database when using the psycopg2 backend,
                 path to the database location when using sqlite3.
```

Reply:
you'll have to quote all references to `config` or `pkgs`, but that's probably better anyway since it doesn't give the impression of some values being constant when they're actually dependent. as a shortcut to generate the proper strings for options in submodules the `options` gained a `__toString` as well (many modules define `opts = options.${module}` similar to `cfg` for convenience):
```patch
@@ -490,6 +492,12 @@ in {
                 sqlite3 = "${cfg.dataDir}/homeserver.db";
                 psycopg2 = "matrix-synapse";
               }.${cfg.settings.database.name};
+              defaultText = literalExpression ''
+              {
+                sqlite3 = "''${options.matrix-synapse.dataDir}/homeserver.db";
+                psycopg2 = "matrix-synapse";
+              }.''${options.matrix-synapse.settings}.database.name;
+              '';
               description = ''
                 Name of the database when using the psycopg2 backend,
                 path to the database location when using sqlite3.
```

Reply:
Applied in [`b9a4444` (#158605)](https://github.com/NixOS/nixpkgs/pull/158605/commits/b9a4444ad7b407d1ee9d22ff28185b7e190b6622)

Reply:
Wow, missed your post @pennae. Very informative, thank you.

Reply:
I agree with @pennae's suggestion :)

Reply:
I applied it like that 4 days ago, it is now squashed.